File: Algorithms/Python3/00121-Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = sys.maxsize
        maxProfit = 0

        for price in prices:
            if price < buy:
                buy = price
            maxProfit = max(maxProfit, price - buy)
        
        return maxProfit

'''
tc: O(n)
buy low, sell high
    - find a day, where price is lowest
    - sell on same day or some day after

keep track of lowest buy price
iterate through prices to find either a) lower buy price or 2) max sell price

'''

File: Algorithms/C++/00121-Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy{0};
        int sell{0};
        int profit{0};
        while(sell < prices.size()){
            if(prices[sell] < prices[buy]){
                buy = sell;
            }
            profit = std::max(profit, prices[sell] - prices[buy]);
            sell++;
        }
        return profit;
    }
};

/*
    tc: O(n)
    sc: O(1)

    2 pointers

    [7,1,5,3,6,4]

    is next number a better buy than current?
        if yes, then move to better buy
        if no, then calculate profit
    return max profit

    profit = 5
    




    


*/